### ISSUE TITLE
Bring the libspiro library directly into FF

### ISSUE THREAD

**Issue by davelab6** [fontforge/fontforge#178]
The libspiro.sf.net library should be be included in FF directly so it becomes a standard feature.

**Comment by JoesCat** [comment]
Creates more "unnecessary" direct work for FontForge developers instead of letting another group take care of libspiro. FontForge is already huge and hard to accept by various distros, so let's not make it even larger and harder to accept/maintain.
See https://bugs.gentoo.org/show_bug.cgi?id=528422#c3 "I withdraw my offer..."
Even though this version of libspiro resides on the FontForge github, it is also usable by GEGL, and potentially be usable by other drawing programs, in which case, those programs would have an interest in helping keep/maintain libspiro in working order.
One of the strengths of linux worth keeping is re-usability of libraries if we can do it.
I would recommend keeping libspiro separate as it currently is right now.

**Comment by mskala** [comment]
On Sun, 25 Jan 2015, Joe wrote:
> Creates more "unnecessary" direct work for FontForge developers instead of
> letting another group take care of libspiro. FontForge is already huge and
> hard to accept by various distros, so let's not make it even larger and
> harder to accept/maintain.

Having separate, and required, libraries makes it harder to compile and
introduces the possibility for library mismatches, so it's not a gimmie
that having them separate necessarily makes it easier to maintain and
distribute.  I don't know which approach is really better.

-- 
Matthew Skala
mskala@ansuz.sooke.bc.ca                 People before principles.
http://ansuz.sooke.bc.ca/

**Comment by JoesCat** [comment]
On January 25, 2015 02:18:19 AM Matthew Skala wrote:
> Having separate, and required, libraries makes it harder to compile and
> introduces the possibility for library mismatches, so it's not a gimmie
> that having them separate necessarily makes it easier to maintain and
> distribute.  I don't know which approach is really better.

From 20120731 we lost the ability to link to external libraries such as 
libspiro and libuninameslist, and we've added a couple more libraries and 
requirements like libreadlines, libzmq, and a few others.
Do you want to merge those too?
or let the expertise remain with those library developers?

I recommend, leave them for those experts to look at.
They are better maintained by people that are more expert in readline, or 
libzmq, or python, or......whatever.
Most patches coming in for FontForge are from developers interested in 
fonts, and are trying to fix FontForge itself, so why are we wanting to dump 
more straw on the camel and duplicate efforts in parallel when you can 
concentrate the expertise where it is best applied.

Libspiro got pulled into FontForge and other projects early and hasn't 
really had a chance to really get applied as a reusable library.
I recommend keep it separate, and let other graphic programs have a chance 
to make use of it.
Grin and bear it for now.
Give it some time.

**Comment by davelab6** [comment]
Who else is using libspiro as a library?

**Comment by adrientetar** [comment]
> Having separate, and required, libraries

And we have much more libraries, and many software use libraries too. I can't see the argument here tbh.

**Comment by adrientetar** [comment]
BTW, spiro is imo superfluous in type design. And it produces curves with a lot of points you have to simplify later.

**Comment by JoesCat** [comment]
On January 25, 2015 12:52:41 PM Dave Crossland wrote:
> Who else is using libspiro as a library?

Right now, it's merged into InkScape.
GIMP might have it merged in also.
(these are examples where libspiro was seen as useful, but a bit buggy, so 
it didn't have much chance to get traction. Even FontForge suffered from 
plenty of crashes when doing spiros (remember 20120731 ?).

GEGL had to drop it due to bugs, and I suspect crashes too, but it appears 
to be getting re-introduced in some distros again. Follow these three bugs, 
and you'll see mentions of GEGL in archlinux, freebsd and gentoo
	https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=196780
	https://bugs.archlinux.org/task/43373
	https://bugs.gentoo.org/show_bug.cgi?id=498410
I haven't followed-up on Fedora, but think it's there too.

It has potential usability for vector graphic programs.
Put it under a similar category like nobody had smart phones 10 years ago.

**Comment by davelab6** [comment]
"merged into" as in what you are recommending Fontforge not do?

**Comment by JoesCat** [comment]
On January 25, 2015 01:32:47 PM Dave Crossland wrote:
> "merged into" as in what you are recommending Fontforge not do?

Yes.
Give the fixed-up version of libspiro a chance to gain traction.

I recommend lightening the load on FontForge "Developers".
Everyone has X hours of time to use.
Throwing more toolkits and libraries into FontForge is not the direction to 
go. It pushes the developers to dabble at everything and be master of 
nothing. Let them put their X amount of time into what they want to do in 
the first place, improve FontForge, not upkeep and maintain parallel 
libraries.

**Comment by davelab6** [comment]
Who are the libspiro developers who are not also FontForge developers? ;)

Its had a 7 year long chance to gain traction, and how many of the 69
changesets actually changed the functionality in a user-visible way? :)

I don't see the reason for maintaining this as a parallel library; let move
it into the fontforge repo tree and remove this as a separately packaged
library.

No one is using it.

**Comment by JoesCat** [comment]
On January 25, 2015 01:57:46 PM Dave Crossland wrote:
> Who are the libspiro developers who are not also FontForge developers?
> ;)

George Williams kept it separate.

Several early patches were scraped-off of the improvements done in Inkscape.
So that would be a big thank you to Shriramana Sharma, and the developer 
that improved the Inkscape version to begin with - I think may have been 
Øyvind Kolås if I read the changelogs correctly.

I haven't touched FontForge since early last year.
The last patches I added to FontForge, I had to start asking for help to 
have patches verified, eg lib readln and a couple new libraries added after 
that. I'd have to upgrade my OS to be able to import some libraries added 
to FontForge since early 2014, but this is unlikely in the near future 
until I unload a big load of TO-DOs off my plate.
Call this count of 4? ...Hopefully this answers the  ;-)

...and if the Originating designer, Raph Levien, himself decides to add 
improvements one day, the door remains open for that possibility. This is 
unlikely to happen for this copy, but much-more unlikely for a fork.



> Its had a 7 year long chance to gain traction, and how many of the 69
> changesets actually changed the functionality in a user-visible way? :)

Will be 70 for 0.3.20150xxx, and your expecting immediate results from a 
community that isn't getting paid to produce results immediately.
7 years is not unreasonable.

1)The robustness patches scraped back into upstream from Inkscape fork.
2)Finite checks added by Shriramana and converge patch added in 2013 is 
quite important as a bug fix.
3) Additional pass/fail return value added in 2013 is important to users 
since not all spiros can converge. Fixes added to Fontforge wouldn't have 
been necessary if this bugfix existed (FontForge currently contains a work-
around to deal with 20071029 version problem).

...these sort of bugfixes and improvements have only been added in 2013



> I don't see the reason for maintaining this as a parallel library;

Neither do I because it creates continued forking and fragmentation.
Right now "this" libspiro appears to be becoming upstream.



> let move it into the fontforge repo tree and remove this as a separately
> packaged library.

Then it becomes yet-another fork.



> No one is using it.

https://github.com/GNOME/gegl/blob/master/configure.ac == libspiro

http://pkgs.fedoraproject.org/cgit/libspiro.git/

https://www.archlinux.org/packages/extra/i686/libspiro/

http://portsmon.freebsd.org/portoverview.py?category=graphics&portname=libspiro

**Comment by JoesCat** [comment]
Hi Dave,

FYI...Inkscape merged libspiro so that it could access spiro.c directly (so 
that it too could do pass/fail checks), which isn't possible through the 
existing 20071029 interface, plus it also needed access to bezctx.h and 
bezctx_intf.h definitions to pass variables, which isn't defined in 
/usr/include or /usr/local/include. Technically speaking, nobody can really 
develop anything useful with the earlier libspiro unless they create their 
own bezctx*.h files since they weren't included in a usr accessible 
location, and they need to do a bit of error checks because the expected 
entry point lacks a pass/fail return. The reason FontForge worked was 
because it had it's own copies of bezctx*.h stored in the Unicode 
directory. and the reason the current FontForge doesn't crash using the 
existing entry point is due to a fair bit of error checks done on the 
returned values (because the existing code lacks pass/fail) through the 
expected entry points.
...hence...fairly hard for someone to have tried developing anything using 
libspiro for the last 7 years without really understanding how it works and 
what happened when it failed to converge. This is resolved with the simple 
return value pass/fail now, and the example code shown in tests.

1)Is this issue considered resolved/closed == no merge.
or
2)Is this issue still open to the probability of still merging code?

We are approaching 2015jan31 in which I was preparing to post version 0.3
For 1, I'll continue as had planned (post update to fontforge/libspiro),
For 2, I'll need to move libspiro elsewhere since merging into FontForge 
makes this copy yet another fork.

**Comment by davelab6** [comment]
On 28 January 2015 at 02:53, Joe <notifications@github.com> wrote:

> We are approaching 2015jan31 in which I was preparing to post version 0.3
> For 1, I'll continue as had planned (post update to fontforge/libspiro),
>

Okay, if you think that would be best... but I think bundling this with
FontForge and moving your work there would be good.

Could we duplicate the libspiro into fontforge and keep them in sync? That
way Spiro is 'a part of' fontforge and will always be available to users.
That's my goal.

**Comment by adrientetar** [comment]
We already need to clone gnulib on every build, I don't see how pulling libspiro would change anything.

**Comment by skef** [comment]
The pattern established since has been working fine. The project does the main mac and windows builds and the distros don't seem unhappy with the libspiro dependency.

**Comment by JoesCat** [comment]
IPE also added the libspiro package too.
